core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategy.java

```diff
@@ -44,51 +45,53 @@
      */
     static RetryStrategy<HttpRequest, HttpResponse> onServerErrorStatus() {
         return onServerErrorStatus(defaultBackoff);
     }
 
     /**
      * Returns the {@link RetryStrategy} that retries the request with the specified {@code backoff}
-     * when the response matches {@link HttpStatusClass#SERVER_ERROR}.
+     * when the response matches {@link HttpStatusClass#SERVER_ERROR} or an {@link Exception} is raised.
      */
     static RetryStrategy<HttpRequest, HttpResponse> onServerErrorStatus(Backoff backoff) {
         requireNonNull(backoff, "backoff");
-        return onStatus(status -> status.codeClass() == HttpStatusClass.SERVER_ERROR ? Optional.of(backoff)
-                                                                                     : Optional.empty());
+        return onStatus((status, thrown) -> {
+            if (thrown != null || (status != null && status.codeClass() == HttpStatusClass.SERVER_ERROR)) {
+                return backoff;
+            }
+            return null;
+        });
     }
 
     /**
-     * Returns the {@link RetryStrategy} that decides to retry the request using
-     * {@code statusBasedBackoffFunction}.
+     * Returns the {@link RetryStrategy} that decides to retry the request using the specified
+     * {@code backoffFunction}.
      *
-     * @param statusBasedBackoffFunction the {@link Function} that returns {@code Optional<Backoff>} according
-     *                                   to the {@link HttpStatus}
+     * @param backoffFunction the {@link BiFunction} that returns the {@link Backoff} according to the
+     *                        {@link HttpStatus} or {@link Throwable}
      */
     static RetryStrategy<HttpRequest, HttpResponse> onStatus(
-            Function<HttpStatus, Optional<Backoff>> statusBasedBackoffFunction) {
-        return new HttpStatusBasedRetryStrategy(statusBasedBackoffFunction);
+            BiFunction<HttpStatus, Throwable, Backoff> backoffFunction) {
+        return new HttpStatusBasedRetryStrategy(backoffFunction);
     }
 
     /**
-     * Returns {@link CompletableFuture} with {@link Optional} that contains {@link Backoff} which will be
-     * used for retry. If the condition does not match, this needs to return {@link Optional#empty()}
-     * to stop retry attempt. Note that {@link ResponseTimeoutException} is not retriable for
-     * the whole retry, but each attempt.
+     * Returns a {@link CompletableFuture} that contains {@link Backoff} which will be used for retry.
+     * If the condition does not match, this will return {@code null} to stop retry attempt.
+     * Note that {@link ResponseTimeoutException} is not retriable for the whole retry, but each attempt.
      *
      * @see RetryingClientBuilder#responseTimeoutMillisForEachAttempt(long)
+     *
+     * @see <a href=https://line.github.io/armeria/advanced-retry.html#per-attempt-timeout>
+     *     Per-attempt-timeout</a>
```

Comment:
- There should be no whitespace between `>` and `Per`, because otherwise it will look like "[ this](https://line.github.io/armeria/)"
- `href` attribute value should be quoted.